File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/net/NetworkManager.java
package app.zxtune.net;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import androidx.annotation.Nullable;

import java.lang.ref.WeakReference;
import java.util.ArrayList;

public class NetworkManager {

  public interface Callback {
    void onNetworkChange(boolean isAvailable);
  }

  @Nullable
  private static NetworkManager instance;

  private final CompositeCallback callbacks;
  private final NetworkStatusReceiver receiver;

  private NetworkManager(Context ctx) {
    this.callbacks = new CompositeCallback();
    this.receiver = new NetworkStatusReceiver(ctx, callbacks);
    ctx.registerReceiver(receiver,
        new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
  }

  public static void initialize(Context ctx) {
    instance = new NetworkManager(ctx);
  }

  public static NetworkManager getInstance() {
    return instance;
  }

  public final boolean isNetworkAvailable() {
    return receiver.isNetworkAvailable();
  }

  public final void subscribe(Callback cb) {
    callbacks.add(cb);
  }

  private static class CompositeCallback implements Callback {

    private final ArrayList<WeakReference<Callback>> delegates = new ArrayList<>(5);

    synchronized final void add(Callback cb) {
      delegates.add(new WeakReference<>(cb));
    }

    @Override
    synchronized public void onNetworkChange(boolean isActive) {
      for (int idx = 0, lim = delegates.size(); idx < lim; ) {
        final Callback cb = delegates.get(idx).get();
        if (cb != null) {
          cb.onNetworkChange(isActive);
          ++idx;
        } else {
          delegates.remove(idx);
          --lim;
        }
      }
    }
  }

  private static class NetworkStatusReceiver extends BroadcastReceiver {

    private final Callback cb;
    private boolean isNetworkAvailable;

    NetworkStatusReceiver(Context ctx, Callback cb) {
      this.cb = cb;
      this.isNetworkAvailable = isNetworkAvailable(ctx);
    }

    final boolean isNetworkAvailable() {
      return isNetworkAvailable;
    }

    @Override
    public void onReceive(Context context, Intent intent) {
      if (ConnectivityManager.CONNECTIVITY_ACTION.equals(intent.getAction())) {
        final boolean newState = isNetworkAvailable(context);
        if (newState != isNetworkAvailable) {
          cb.onNetworkChange(isNetworkAvailable = newState);
        }
      }
    }

    private static boolean isNetworkAvailable(Context ctx) {
      final ConnectivityManager connectivityManager =
          (ConnectivityManager) ctx.getSystemService(Context.CONNECTIVITY_SERVICE);
      final NetworkInfo info = connectivityManager != null ? connectivityManager.getActiveNetworkInfo() : null;
      return info != null && info.isConnected();
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/net/NetworkManager.kt
package app.zxtune.net

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.annotation.VisibleForTesting
import androidx.startup.AppInitializer
import androidx.startup.Initializer
import app.zxtune.Logger
import app.zxtune.Releaseable
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.stateIn

private val LOG = Logger(NetworkManager::class.java.name)

class NetworkManager @VisibleForTesting constructor(
    src: NetworkStateSource, dispatcher: CoroutineDispatcher
) {

    val networkAvailable = callbackFlow {
        LOG.d { "Start monitoring" }
        val handle = src.monitorChanges {
            LOG.d { "NetworkState=${it}" }
            trySend(it)
        }
        awaitClose {
            handle.release()
        }
        LOG.d { "End monitoring" }
    }.stateIn(CoroutineScope(dispatcher), SharingStarted.Eagerly, src.isNetworkAvailable())

    init {
        LOG.d { "Created" }
    }

    companion object {
        fun from(ctx: Context) =
            AppInitializer.getInstance(ctx).initializeComponent(Factory::class.java)
    }

    class Factory : Initializer<NetworkManager> {
        override fun create(ctx: Context) = NetworkManager(
            NetworkStateSource.create(ctx), Dispatchers.Main.immediate
        )

        override fun dependencies(): List<Class<out Initializer<*>>> = emptyList()
    }
}

typealias NetworkStateCallback = (Boolean) -> Unit

@VisibleForTesting
interface NetworkStateSource {
    fun isNetworkAvailable(): Boolean
    fun monitorChanges(cb: NetworkStateCallback): Releaseable

    companion object {
        fun create(ctx: Context) = if (Build.VERSION.SDK_INT >= 24) {
            NetworkStateApi24(ctx)
        } else {
            NetworkStatePre24(ctx)
        }
    }
}

// Based on solutions from
// https://stackoverflow.com/questions/36421930/connectivitymanager-connectivity-action-deprecated
@RequiresApi(24)
private class NetworkStateApi24(ctx: Context) : NetworkStateSource {

    private val manager = ctx.getConnectivityManager()

    override fun isNetworkAvailable() = manager?.activeNetwork?.let { network ->
        manager.getNetworkCapabilities(network)?.run {
            hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) && hasCapability(
                NetworkCapabilities.NET_CAPABILITY_VALIDATED
            )
        }
    } ?: false

    override fun monitorChanges(cb: NetworkStateCallback): Releaseable {
        val networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) = cb.invoke(true)
            override fun onLost(network: Network) = cb.invoke(false)
        }
        manager?.registerDefaultNetworkCallback(networkCallback)
        return Releaseable { manager?.unregisterNetworkCallback(networkCallback) }
    }
}

@Suppress("DEPRECATION")
private class NetworkStatePre24(private val ctx: Context) : NetworkStateSource {

    override fun isNetworkAvailable() =
        true == ctx.getConnectivityManager()?.activeNetworkInfo?.isConnected

    override fun monitorChanges(cb: NetworkStateCallback): Releaseable {
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                val isNoConnectivity =
                    intent?.extras?.getBoolean(ConnectivityManager.EXTRA_NO_CONNECTIVITY) ?: true
                cb.invoke(!isNoConnectivity)
            }
        }
        ctx.registerReceiver(receiver, IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION))
        return Releaseable { ctx.unregisterReceiver(receiver) }
    }
}

private fun Context.getConnectivityManager() =
    getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
